src/app/chat/chat.service.ts

```diff
@@ -178,4 +186,203 @@ export class ChatService {
   private getRamdomNumber() {
     return Math.floor(Math.random() * 19) + 1;
   }
+
+  getDate(prams: getDatePrams) {
+    const months = [
+      "Jan",
+      "Feb",
+      "Mar",
+      "Apr",
+      "May",
+      "Jun",
+      "Jul",
+      "Aug",
+      "Sept",
+      "Oct",
+      "Nov",
+      "Dec"
+    ];
+    let returnDate = "";
+    let returnTime = "";
+    if (prams.date) {
+      var UTCDate = prams.date.replace(/-/g, "/") + " UTC";
+      var chatDate = new Date(UTCDate);
+      var today = new Date();
+      var diff = today.getDate() - chatDate.getDate();
+      if (diff === 0) {
+        returnDate = "Today";
+      } else if (diff === 1) {
+        returnDate = "Yesterday";
+      } else {
+        returnDate =
+          months[chatDate.getMonth()] + " " + chatDate.getDate().toString();
+      }
+
+      returnTime = chatDate.toLocaleString("en-US", {
+        hour: "numeric",
+        minute: "numeric",
+        hour12: true
+      });
+
+      switch (prams.type) {
+        case "list":
+          if (diff === 0) {
+            return returnTime;
+          } else {
+            return returnDate;
+          }
+        case "room":
+          if (prams.messageList && prams.messageIndex) {
+            if (prams.messageList[prams.messageIndex - 1]) {
+              var currentMessageTime = new Date(
+                prams.messageList[prams.messageIndex].sent_time
+              );
+              var oldMessageTime = new Date(
+                prams.messageList[prams.messageIndex - 1].sent_time
+              );
+              var dateDiff =
+                currentMessageTime.getDate() - oldMessageTime.getDate();
+              if (dateDiff === 0) {
+                return returnTime;
+              } else {
+                return returnDate + " " + returnTime;
+              }
+            }
+          }
+      }
+    } else {
+      return "";
+    }
+  }
+
+  /**
+   * modify the response
+   *  - set chat avatar color
+   * @param {Array} response
+   */
+  private _normaliseChatListResponse(response): Array<any> {
+    if (response) {
+      this.chatList = [];
+      this.chatColors = this.storage.get("chatAvatarColors");
+      if (!this.chatColors) {
+        return this.setChatAvatarColors(response, null, "nocolor");
+      } else {
+        return this.setChatAvatarColors(response, this.chatColors, "havecolor");
+      }
+    }
+  }
+
+  /**
+   * this method check old avatar colors and set the releted one.
+   * if there no old avatar colors it will create new color and add that to service variable.
+   * @param {Array} response
+   * @param {Array} chatColors
+   * @param {String} status
+   */
+  private setChatAvatarColors(response, chatColors, status): Array<any> {
+    let index = 0;
+    for (index = 0; index < response.length; index++) {
```

Comment:
Array.forEach() run in synchronously so because of that to execute some function or code after for each we have track it's index using variables or we have use promises to check is loop is over. and also Array.forEach is bit more costly when compair with for loops. in here i use for loop it's easy to get to know loop is over and execute function after loop over.

Reply:
Firstly, Array.forEach() is not asynchronous, so don't need to use promise for that. Try this:
```
let b = [1,2,3,4,5,6];
b.forEach((v, i) => {
  b[i]++;
});
console.log(b);
```

Secondly, regarding the performance of `for` and `forEach()`, `for` is faster than `forEach()`, but the difference is too little that we can just omit. There're benefits of using `forEach()` as well, such as readability.

So, the reason I pointed this out is that I think it's better to make it consistent among the codebase. Since we use `forEach()` in all other places, it's better to use it here as well.

Anyway, there's no much difference regarding the functionality.

Reply:
yes sure. I will change for to foreach.